### PR TITLE
Improve schema inference cache, respect format settings that can change the schema

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve schema inference cache, respect format settings that can change the schema.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Processors/Formats/Impl/JSONColumnsBlockInputFormatBase.cpp [hunk 1]
```diff
@@ -214,8 +218,15 @@ NamesAndTypesList JSONColumnsSchemaReaderBase::readSchema()
         do
         {
             auto column_name_opt = reader->readColumnStart();
-            /// If format doesn't have named for columns, use default names 'c1', 'c2', ...
-            String column_name = column_name_opt.has_value() ? *column_name_opt : "c" + std::to_string(iteration + 1);
+            /// If format doesn't have named for columns, use names from setting column_names_for_schema_inference or default names 'c1', 'c2', ...
```
Comment (kssenii) [inline src/Processors/Formats/Impl/JSONColumnsBlockInputFormatBase.cpp:221]:
```suggestion
            /// If format doesn't have names for columns, use names from setting column_names_for_schema_inference or default names 'c1', 'c2', ...
```

File: src/Processors/Formats/Impl/TemplateRowInputFormat.cpp [hunk 2]
```diff
@@ -566,13 +566,32 @@ void registerTemplateSchemaReader(FormatFactory & factory)
...
+        {
+            size_t index = 0;
+            auto idx_getter = [&](const String &) -> std::optional<size_t> { return index++; };
+            auto row_format = fillRowFormat(settings, idx_getter, false);
+            std::unordered_set<FormatSettings::EscapingRule> visited_escaping_rules;
+            String result = fmt::format("row_format={}, resultset_format={}, row_between_delimiter={}",
+                settings.template_settings.row_format,
+                settings.template_settings.resultset_format,
+                settings.template_settings.row_between_delimiter);
+            for (auto escaping_rule : row_format.escaping_rules)
```
Comment (kssenii) [inline src/Processors/Formats/Impl/TemplateRowInputFormat.cpp:587]: `const auto &`?

**Review by kssenii — APPROVED** [review]

File: src/Processors/Formats/Impl/TemplateRowInputFormat.cpp (hunk 2 above)
Comment (Avogar) [reply src/Processors/Formats/Impl/TemplateRowInputFormat.cpp:587]: `escaping_rule ` is a Enum value, using const reference doesn't make sense, copying it as cheap as creating a reference

File: src/Processors/Formats/Impl/JSONColumnsBlockInputFormatBase.cpp (hunk 1 above)
Comment (Avogar) [reply src/Processors/Formats/Impl/JSONColumnsBlockInputFormatBase.cpp]: Thanks, fixed

**Pull request merged**